Reduce memory usage of flag timeseries assets by ~half

zschira opened the pull request:
# Overview

Closes #4678.

## What problem does this address?
This PR improves the memory performance of the `_flag_timeseries` assets that make up one portion of our demand imputation pipeline that operates on FERC 714 and EIA 930 data. The specific assets impacted by this change are:

- `_out_ferc714__hourly_planning_area_demand_timeseries_matrix_flags`
- `_out_eia930__combined_demand_timeseries_matrix_flags`

## What did you change?
The peak memory usage comes from a specific call to `numpy.nanmedian` that operates on a very large 3D array. To decrease memory usage I've brought `Dask` back in to our dependency stack (after a short lived absence) due to their [array api](https://docs.dask.org/en/stable/array.html), which implements much of the Numpy api, but allows out-of-memory processing.

For more details on other options explored see #4678.

### Results
This change reduces peak memory usage by roughly half. This still leads to quite high memory usage for the 714 asset (11.9GB), but it is still a significant improvement. I spent a good amount of time attempting to tweak Dask configuration and [array chunksize](https://docs.dask.org/en/stable/array-chunks.html#chunks) to try to further decrease emory usage, but this is as low as I could get it. There may be further room for improvement here, but median computation is inherently expensive, so I'm not sure there's a lot more to squeeze out of this.

#### `_out_ferc714__hourly_planning_area_demand_timeseries_matrix_flags`
[714-baseline-profile.html](https://github.com/user-attachments/files/23528501/memray-flamegraph-_out_ferc714__hourly_planning_area_demand_timeseries_matrix_flags-2025-11-11T19.07.21.html) -> [714-optimized-profile.html](https://github.com/user-attachments/files/23528810/memray-flamegraph-_out_ferc714__hourly_planning_area_demand_timeseries_matrix_flags-2025-11-11T18.00.36.html)

#### `_out_eia930__combined_demand_timeseries_matrix_flags`
[930-basline-profile.html](https://github.com/user-attachments/files/23528825/memray-flamegraph-_out_eia930__combined_imputed_demand_timeseries_matrix_flags-2025-11-10T19.03.21.html) -> [930-optimized-profile.html](https://github.com/user-attachments/files/23528846/memray-flamegraph-_out_eia930__combined_imputed_demand_timeseries_matrix_flags-2025-11-11T14.49.28.html)

# Testing
To validate that the results of this change I ran the `check_against_nightly` script on the following assets:

- `out_ferc714__hourly_planning_area_demand`
- `out_eia930__hourly_subregion_demand`
- `out_eia930__hourly_operations`

These assets are all downstream of the asset actually being modified, but they contain the results of the imputation. Because the actual imputation is non-deterministic, this ends up failing, but for all of the tables I only see differences in the `demand_imputed_pudl_mwh` column. Since this change applies to the flagging step, we would expect to see changes in the `demand_imputed_pudl_mwh_imputation_code` column if this change had impacted flagging.

## To-do list

- [x] Figure out why 714 asset has different row counts
    - I had an outdated upstream asset causing the mismatch.

zaneselvans commented on src/pudl/analysis/timeseries_cleaning.py:
```diff
@@ -878,11 +881,20 @@ def median_of_rolling_median_offset(
+                    np.load(array_file, mmap_mode="r"), chunks=(21, 10791, 19)
+                ),
+                axis=0,
+            ).compute()
```
I wonder if this could also be done in a memory efficient way with DuckDB, which is apparently happy to [query numpy arrays](https://duckdb.org/docs/stable/guides/python/import_numpy)?

zschira replied:
Wow duckdb can do everything. Unfortunately, I spent some time investigating this option this afternoon and ran into the same issues as polars where you need to offload the array onto disk to get any efficiency gains, and that ends up being a massive bottleneck. Also duckdb can't do `SELECT median(*) FROM table`, so you need to auto-generate a query to compute the median on each individual column and there's 10s of thousands of columns.

krivard commented:
We might not actually need dask for this -- on a hunch I tried a drop-in replacement for nanmedian from `bottleneck` and got a 2/3 reduction in memory (24->8) and 1/3 reduction in running time (18->12)

* profile from main: 
[714-baseline-profile.html](https://github.com/user-attachments/files/23550404/memray-flamegraph-_out_ferc714__hourly_planning_area_demand_timeseries_matrix_flags-2025-11-14T14.46.53.html)

* profile with the below diff: 
[714-bottleneck-profile.html](https://github.com/user-attachments/files/23550406/memray-flamegraph-_out_ferc714__hourly_planning_area_demand_timeseries_matrix_flags-2025-11-14T15.09.25.html)

```
diff --git a/src/pudl/analysis/timeseries_cleaning.py b/src/pudl/analysis/timeseries_cleaning.py
index 7e262b2c5..d300401b7 100644
--- a/src/pudl/analysis/timeseries_cleaning.py
+++ b/src/pudl/analysis/timeseries_cleaning.py
@@ -37,6 +37,7 @@ from collections.abc import Callable, Sequence
 from dataclasses import dataclass, field
 from typing import Any, Literal
 
+import bottleneck as bn
 import numpy as np
 import pandas as pd
 import pandera.pandas as pa
@@ -882,7 +883,7 @@ def median_of_rolling_median_offset(
         warnings.filterwarnings(
             "ignore", category=RuntimeWarning, message="All-NaN slice encountered"
         )
-        return np.nanmedian(shifted, axis=0)
+        return bn.nanmedian(shifted, axis=0)
 
 
 def rolling_iqr_of_rolling_median_offset(
```

zschira commented:
@krivard nice! I wasn't familiar with `bottleneck`, but it's simpler and more efficient than `dask`, so I've switched over to this implementation.

Also removed the dask dependency.

zaneselvans commented on src/pudl/analysis/timeseries_cleaning.py:
```diff
@@ -882,7 +883,7 @@ def median_of_rolling_median_offset(
             "ignore", category=RuntimeWarning, message="All-NaN slice encountered"
         )
-        return np.nanmedian(shifted, axis=0)
+        return bn.nanmedian(shifted, axis=0)
```
It will probably be hard to beat this for "memory reduction per characters changed"

zaneselvans commented:
Imma go ahead and merge this guy in since it's such a tiny change for such a big improvement.
